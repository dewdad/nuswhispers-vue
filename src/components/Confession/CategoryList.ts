import Vue from 'vue';
import Component from 'vue-class-component';
import { Category } from '../../store/state';

@Component({
  props: {
    ids: Array,
  },
})
export default class CategoryList extends Vue {
  public ids: string[];

  get categories(): Category[] {
    return this.$store.getters.categoriesByIds(this.ids);
  }
}
