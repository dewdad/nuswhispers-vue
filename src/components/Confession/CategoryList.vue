<template>
  <div v-if="categories.length">
    <template v-for="(category, idx) in categories">
      <router-link :to="'/category/' + category.confessionCategoryId">{{ category.confessionCategory }}</router-link><span v-if="idx + 1 < categories.length">, </span>
    </template>
  </div>
  <div v-else>
    No Category
  </div>
</template>

<script lang="ts" src="./CategoryList.ts" />
