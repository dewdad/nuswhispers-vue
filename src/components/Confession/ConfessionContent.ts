import flow from 'lodash/flow';
import Vue from 'vue';
import Component from 'vue-class-component';
import escapeHtml from '../../utils/escape-html';

const linkTags = (content: string): string => {
  const tags = content.split(/(#\w+)/);
  let formatted = '';

  tags.map((tag) => {
    formatted += /(#\w+)/.test(tag) ?
      `<a href="/tag/${tag.substring(1)}">${tag}</a>` :
      tag;
  });

  return formatted;
};

const linkUrls = (content: string): string => {
  const regex = /(\b(https?|ftp):\/\/[A-Z0-9+&@#\/%?=~_|!:,.;-]*[-A-Z0-9+&@#\/%=~_|])/gim;
  if (content.match(regex)) {
    content = content.replace(regex, '<a rel=\"noopener noreferrer\" href=\"$1\" target=\"_blank\">$1</a>');
  }
  return content;
};

@Component({
  props: {
    content: String,
  },
})
export default class ConfessionContent extends Vue {
  public content: string;

  get formattedContent(): string {
    return flow(escapeHtml, linkTags, linkUrls)(this.content);
  }
}
