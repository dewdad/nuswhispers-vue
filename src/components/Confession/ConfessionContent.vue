<template>
  <p v-html="formattedContent" />
</template>

<script lang="ts" src="./ConfessionContent.ts" />
