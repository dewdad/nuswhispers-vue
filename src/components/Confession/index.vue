<template>
  <article :class="$style.confession">
    <box>
      <header :class="$style.header">
        <category-list :class="$style.categories" :ids="item.categories" />
        <div :class="$style.social">
          <a target="_blank" rel="nofollow norel" :href="'https://twitter.com/intent/tweet?url=http%3A%2F%2Fnuswhispers.com%2Fconfession%2F' + item.confessionId">
            <icon :class="$style.actionIcon" id="social-twitter" />
          </a>
          <a target="_blank" rel="nofollow noopener" :href="'https://www.facebook.com/dialog/share?app_id=1577825682475577&amp;type=dialog&amp;href=http%3A%2F%2Fnuswhispers.com%2Fconfession%2F' + item.confessionId">
            <icon :class="$style.actionIcon" id="social-facebook" />
          </a>
        </div>
      </header>

      <div :class="$style.content">
        <router-link :to="'/confession/' + item.confessionId">#{{ item.confessionId }}:</router-link>
        <confession-content :class="$style.contentInject" :content="item.content" />
        <img :class="$style.contentImage" v-if="item.images" :src="item.images" :alt="'Confession #' + item.confessionId" />
        <p>
          <a target="_blank" rel="nofollow noopener" :href="fbLink">View on Facebook</a>
        </p>
      </div>

      <footer :class="$style.footer">
        <div :class="$style.timestamp">
          <icon :class="$style.footerIcon" id="time" />
          {{ humanCreatedAt }}
        </div>
        <div :class="$style.actions">
          <a :href="fbLink" :class="$style.action">
            <icon :class="$style.footerIcon" id="message" />
            {{ item.fbCommentCount }}
          </a>
          <a :href="fbLink" :class="$style.action">
            <icon :class="$style.footerIcon" id="thumbs-up" />
            {{ item.fbLikeCount }}
          </a>
          <!--<a href="#" :class="$style.action">
            <icon :class="$style.footerIcon" id="bookmark" />
            0
          </a>-->
        </div>
      </footer>
    </box>
  </article>
</template>

<style module>
.confession {
  margin-bottom: 1rem;
}

.header {
  border-bottom: 1px solid #eee;
  color: #999;
  display: flex;
  padding-bottom: .5rem;
  margin-bottom: 1rem;
}

.categories {
  flex: 2;
}

.social {
  text-align: right;
}

.icon {
  fill: #999;
  height: 24px;
  transition: fill .25s;
  width: 24px;
  vertical-align: middle;
}

.actionIcon {
  composes: icon;
  margin-left: .5rem;
}

.actionIcon:hover {
  fill: #333;
}

.content {
  line-height: 1.5rem;
}

.contentInject {
  display: inline;
  white-space: pre-wrap;
  word-break: break-word;
}

.contentImage {
  display: block;
  height: auto;
  max-width: 100%;
  padding-top: 1rem;
}

.link {
  text-align: right;
}

.footer {
  border-top: 1px solid #eee;
  color: #999;
  display: flex;
  padding-top: .5rem;
  margin-top: 1rem;
}

.timestamp {
  margin-left: -5px;
}

.footerIcon {
  composes: icon;
  fill: #ccc;
  vertical-align: -25%;
}

.actions {
  flex: 2;
  text-align: right;
}

.action {
  color: #999;
  fill: #999;
  margin-left: .5rem;
  text-decoration: none;
}

.action:hover {
  color: #333;
  fill: #333;
  text-decoration: none;
}
</style>

<script lang="ts" src="./index.ts" />
