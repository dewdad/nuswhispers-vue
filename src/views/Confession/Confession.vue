<template>
  <div class="single">
    <confession :item="item" />
  </div>
</template>

<script lang="ts" src="./Confession.ts" />
