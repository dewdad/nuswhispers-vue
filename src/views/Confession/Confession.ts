import truncate from 'lodash/truncate';
import Vue from 'vue';
import { Route } from 'vue-router';
import ASyncComponent, { InitialDataProps } from '../../components/ASyncComponent';
import Confession from '../../components/Confession/index.vue';
import meta from '../../mixins/meta';
import title from '../../mixins/title';
import { Store } from '../../store';
import { Confession as ConfessionState } from '../../store/state';
import * as metaUtils from '../../utils/meta';

@ASyncComponent({
  components: {
    Confession,
  },
  mixins: [meta, title],
})
export default class ConfessionView extends Vue {
  public loading: boolean = false;

  get item(): ConfessionState {
    return this.$store.getters.listById(this.$route.params.id);
  }

  public beforeRouteUpdate(to: Route, from: Route, next: () => void) {
    // Fetch data if category ID is different from the previous one.
    if (to.params.id !== from.params.id) {
      return this.fetchInitialData({ store: this.$store, route: to }).then(() => next());
    }
    next();
  }

  public fetchInitialData({ store, route }: InitialDataProps): Promise<any> {
    this.loading = true;
    return store.dispatch('getById', route.params.id).then(() => this.loading = false);
  }

  public title(): string {
    return `#${this.$route.params.id} - NUSWhispers`;
  }

  public meta(): Array<{ [key: string]: string }> {
    const imageTags = this.item.images ?
        metaUtils.generateImage(this.item.images) :
        metaUtils.defaultImage();

    return [
      ...metaUtils.generateTitle(`#${this.$route.params.id} - NUSWhispers`),
      ...metaUtils.generateDescription(truncate(this.item.content, { length: 240 })),
      ...imageTags,
      {
        content: this.item.images ? 'summary_large_image' : 'summary',
        property: 'twitter:card',
      },
    ];
  }
}
