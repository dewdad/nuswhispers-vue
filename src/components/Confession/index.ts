import distanceInWordsStrict from 'date-fns/distance_in_words_strict';
import parse from 'date-fns/parse';
import Vue from 'vue';
import Component from 'vue-class-component';
import { Confession } from '../../store/state';
import CategoryList from './CategoryList';
import ConfessionContent from './ConfessionContent';

@Component({
  components: {
    CategoryList,
    ConfessionContent,
  },
  name: 'confession',
  props: {
    item: Object,
  },
})
export default class ConfessionComponent extends Vue {
  public item: Confession;

  get fbLink(): string {
    return `https://www.facebook.com/nuswhispers/posts/${this.item.fbPostId}`;
  }

  get humanCreatedAt(): string {
    // Pad a 'Z' at the end to set it as UTC (which is returned from the API).
    return distanceInWordsStrict(new Date(), parse(this.item.createdAt + 'Z'), { addSuffix: true });
  }
}
