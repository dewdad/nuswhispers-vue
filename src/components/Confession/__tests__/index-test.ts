import format from 'date-fns/format';
import subDays from 'date-fns/sub_days';
import Vue from 'vue';
import { createRenderer } from 'vue-server-renderer';
import Vuex from 'vuex';
import createStore from '../../../store';
import Confession from '../index.vue';

Vue.use(Vuex);

// Mock current date to 2017-05-03 00:00:00.
Date.now = jest.fn(() => 1493769600000);

describe('Confession', () => {
  const renderer = createRenderer();

  it('renders correctly', () => {
    const Ctor = Vue.extend(Confession);
    const vm = new Ctor({
      propsData: {
        item: {
          categories: [],
          confessionId: 1,
          content: 'Hello World!',
          createdAt: '2017-05-01 00:00:00',
          facebookInformation: '1',
          favourites: [],
          fbCommentCount: 0,
          fbLikeCount: 0,
          fbPostId: 1,
          images: null,
          status: 'Approved',
          statusUpdatedAt: '2017-05-11 11:15:02',
          statusUpdatedAtTimestamp: '1494501302',
          updatedAt: '2017-05-12 13:01:57',
          views: 10,
        },
      },
      store: createStore(),
    });

    renderer.renderToString(vm, (err: Error | undefined, html: string) => {
      expect(html).toMatchSnapshot();
    });
  });
});
