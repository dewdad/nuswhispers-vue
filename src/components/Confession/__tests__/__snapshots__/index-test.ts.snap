// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confession renders correctly 1`] = `
<article data-server-rendered="true" class="confession">
    <box>
        <header class="header">
            <div class="categories">
                No Category
            </div>
            <div class="social">
                <a target="_blank" rel="nofollow norel" href="https://twitter.com/intent/tweet?url=http%3A%2F%2Fnuswhispers.com%2Fconfession%2F1">
                    <icon id="social-twitter" class="actionIcon"></icon>
                </a>
                <a target="_blank" rel="nofollow noopener" href="https://www.facebook.com/dialog/share?app_id=1577825682475577&amp;type=dialog&amp;href=http%3A%2F%2Fnuswhispers.com%2Fconfession%2F1">
                    <icon id="social-facebook" class="actionIcon"></icon>
                </a>
            </div>
        </header>
        <div class="content">
            <router-link to="/confession/1">#1:</router-link>
            <p class="contentInject">Hello World!</p>
            <!---->
            <p><a target="_blank" rel="nofollow noopener" href="https://www.facebook.com/nuswhispers/posts/1">View on Facebook</a></p>
        </div>
        <footer class="footer">
            <div class="timestamp">
                <icon id="time" class="footerIcon"></icon>
                2 days ago
            </div>
            <div class="actions">
                <a href="https://www.facebook.com/nuswhispers/posts/1" class="action">
                    <icon id="message" class="footerIcon"></icon>
                    0
                </a>
                <a href="https://www.facebook.com/nuswhispers/posts/1" class="action">
                    <icon id="thumbs-up" class="footerIcon"></icon>
                    0
                </a>
            </div>
        </footer>
    </box>
</article>
`;
